### PR TITLE
Propagate symbolic dimensions through idioms like x.view(y.size())

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62914 Add handling of ifs to shape propagation
* #62911 Small shape analysis changes
* #62910 Add a few peepholes
* **#61975 Propagate symbolic dimensions through idioms like x.view(y.size())**

Propagate symbolic dimensions through size calls. We did this by associating SymbolicSizes with integer inputs by looking through their constructors for `x.size(1)` or `x.size()` nodes.

Differential Revision: [D30196948](https://our.internmc.facebook.com/intern/diff/D30196948)

### CODE REVIEW

File: torch/csrc/jit/passes/symbolic_shape_analysis.cpp [hunk 1]
```diff
@@ -109,15 +109,68 @@ struct SymbolicShapeAnalyzer {
...
+            }
+          }
+          node_symbolic_input_indices.emplace_back(
+              i, c10::SymbolicShape(shape));
+        } else if (
+            node_->input(i)->node()->kind() == aten::size &&
+            node_->input(i)->uses().size() == 1) {
+          node_symbolic_input_indices.emplace_back(
+              i,
+              node_->input(i)
```
Comment (eellison) [inline torch/csrc/jit/passes/symbolic_shape_analysis.cpp:168]: damn formatting here kinda sucks..

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: test/jit/test_symbolic_shape_analysis.py [hunk 2]
```diff
@@ -116,3 +116,20 @@ def apply(fn):
...
             self.assertEqual(next(t.graph.outputs()).type().symbolic_sizes(), [4, 4, 8])
+
+    def test_size_and_sizes(self):
+        @torch.jit.script
+        def foo(x, y):
+            return x.view(y.size(0), 8, y.size(-1))
+
+        @torch.jit.script
+        def foo2(x, y):
+            return x.view(y.size())
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/jit/test_symbolic_shape_analysis.py:127]: what do you mean ?

File: torch/csrc/jit/passes/symbolic_shape_analysis.cpp [hunk 3]
```diff
@@ -109,15 +109,60 @@ struct SymbolicShapeAnalyzer {
...
-        if (auto ival = toIValue(node_->input(i))) {
-          replaceWithIValue(graph_->inputs().at(i), *ival);
+      } else if (auto ival = toIValue(node_->input(i))) {
+        replaceWithIValue(graph_->inputs().at(i), *ival);
+      } else if (
+          type->cast<ListType>() &&
+          type->cast<ListType>()->getElementType()->cast<IntType>()) {
+        if (node_->input(i)->node()->kind() == prim::ListConstruct &&
+            node_->input(i)->uses().size() == 1) {
+          // it is a very common in graphs to see patterns like:
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/symbolic_shape_analysis.cpp:125]: lol, yea that would be nice. i haven't really thought about the feasibility of canonicalizing all these ops yet but that would be cool! i think sometimes they have slightly different semantics with wrt/views/copies etc

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30196948).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30196948).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30196948).

**Pull request closed without merging**